homeassistant/components/zha/helpers.py

```diff
@@ -0,0 +1,1201 @@
+        super().__init__()
+        self.device: Device = device
+        self.gateway_proxy: ZHAGatewayProxy = gateway_proxy
+        self._unsubs: list[Callable[[], None]] = []
+        self._unsubs.append(self.device.on_all_events(self._handle_event_protocol))
+
+    @property
+    def device_id(self) -> str:
+        """Return the HA device registry device id."""
+        return self._ha_device_id
+
+    @device_id.setter
+    def device_id(self, device_id: str) -> None:
+        """Set the HA device registry device id."""
+        self._ha_device_id = device_id
+
+    @property
+    def device_info(self) -> dict[str, Any]:
+        """Return a device description for device."""
+        ieee = str(self.device.ieee)
+        time_struct = time.localtime(self.device.last_seen)
+        update_time = time.strftime("%Y-%m-%dT%H:%M:%S", time_struct)
+        return {
+            ATTR_IEEE: ieee,
+            ATTR_NWK: self.device.nwk,
+            ATTR_MANUFACTURER: self.device.manufacturer,
+            ATTR_MODEL: self.device.model,
+            ATTR_NAME: self.device.name or ieee,
+            ATTR_QUIRK_APPLIED: self.device.quirk_applied,
+            ATTR_QUIRK_CLASS: self.device.quirk_class,
+            ATTR_QUIRK_ID: self.device.quirk_id,
+            ATTR_MANUFACTURER_CODE: self.device.manufacturer_code,
+            ATTR_POWER_SOURCE: self.device.power_source,
+            ATTR_LQI: self.device.lqi,
+            ATTR_RSSI: self.device.rssi,
+            ATTR_LAST_SEEN: update_time,
+            ATTR_AVAILABLE: self.device.available,
+            ATTR_DEVICE_TYPE: self.device.device_type,
+            ATTR_SIGNATURE: self.device.zigbee_signature,
+        }
+
+    @property
+    def zha_device_info(self) -> dict[str, Any]:
+        """Get ZHA device information."""
+        device_info: dict[str, Any] = {}
+        device_info.update(self.device_info)
+        device_info[ATTR_ACTIVE_COORDINATOR] = self.device.is_active_coordinator
+        device_info["entities"] = [
+            {
+                "entity_id": entity_ref.ha_entity_id,
+                ATTR_NAME: entity_ref.ha_device_info[ATTR_NAME],
+            }
+            for entity_ref in self.gateway_proxy.ha_entity_refs[self.device.ieee]
+        ]
+
+        topology = self.gateway_proxy.gateway.application_controller.topology
+        device_info[ATTR_NEIGHBORS] = [
+            {
+                "device_type": neighbor.device_type.name,
+                "rx_on_when_idle": neighbor.rx_on_when_idle.name,
+                "relationship": neighbor.relationship.name,
+                "extended_pan_id": str(neighbor.extended_pan_id),
+                "ieee": str(neighbor.ieee),
+                "nwk": str(neighbor.nwk),
+                "permit_joining": neighbor.permit_joining.name,
+                "depth": str(neighbor.depth),
+                "lqi": str(neighbor.lqi),
+            }
+            for neighbor in topology.neighbors[self.device.ieee]
+        ]
```

Comment:
I personally dislike the mix between constants and normal strings

Reply:
updated